File: examples_v2/example_sdk_typescript/_api/_productNode_api.ts
/**
 * API class for ProductNode instances.
 *
 * @packageDocumentation
 */

import { InstanceAPI } from "../instance_api/api.ts";
import type { PygenClientConfig } from "../instance_api/auth/index.ts";
import type { ViewReference, InstanceId } from "../instance_api/types/references.ts";
import type { Aggregation, PropertySort, SortDirection } from "../instance_api/types/query.ts";
import type { AggregateResponse, Page } from "../instance_api/types/responses.ts";
import { InstanceList } from "../instance_api/types/instance.ts";

import {
  PRODUCT_NODE_VIEW,
  ProductNode,
  ProductNodeFilter,
  ProductNodeList,
} from "../data_classes/index.ts";

/**
 * Creates a property reference for sorting.
 *
 * @param viewRef - The view reference
 * @param propertyName - The property name
 * @returns A property path array
 */
function createPropertyRef(
  viewRef: ViewReference,
  propertyName: string,
): [string, string, string] {
  return [viewRef.space, `${viewRef.externalId}/${viewRef.version}`, propertyName];
}

/**
 * API for ProductNode instances with type-safe filter methods.
 */
export class ProductNodeApi extends InstanceAPI<ProductNode> {
  /**
   * Creates a new ProductNodeApi.
   *
   * @param config - Client configuration for API access
   */
  constructor(config: PygenClientConfig) {
    super(config, PRODUCT_NODE_VIEW, "node");
  }

  /**
   * Builds a ProductNodeFilter from the given options.
   *
   * @param options - Filter options
   * @returns A configured ProductNodeFilter
   */
  private _buildFilter(options: {
    name?: string | readonly string[];
    namePrefix?: string;
    description?: string | readonly string[];
    descriptionPrefix?: string;
    minPrice?: number;
    maxPrice?: number;
    minQuantity?: number;
    maxQuantity?: number;
    active?: boolean;
    minCreatedDate?: Date | string;
    maxCreatedDate?: Date | string;
    minUpdatedTimestamp?: Date | string;
    maxUpdatedTimestamp?: Date | string;
    category?: string | InstanceId | readonly [string, string] | readonly (string | InstanceId | readonly [string, string])[];
    externalIdPrefix?: string;
    space?: string | readonly string[];
  }): ProductNodeFilter {
    const filter = new ProductNodeFilter("and");
    filter.name.equalsOrIn(options.name ?? null);
    filter.name.prefix(options.namePrefix ?? null);
    filter.description.equalsOrIn(options.description ?? null);
    filter.description.prefix(options.descriptionPrefix ?? null);
    filter.price.greaterThanOrEquals(options.minPrice ?? null);
    filter.price.lessThanOrEquals(options.maxPrice ?? null);
    filter.quantity.greaterThanOrEquals(options.minQuantity ?? null);
    filter.quantity.lessThanOrEquals(options.maxQuantity ?? null);
    filter.active.equals(options.active ?? null);
    filter.createdDate.greaterThanOrEquals(options.minCreatedDate ?? null);
    filter.createdDate.lessThanOrEquals(options.maxCreatedDate ?? null);
    filter.updatedTimestamp.greaterThanOrEquals(options.minUpdatedTimestamp ?? null);
    filter.updatedTimestamp.lessThanOrEquals(options.maxUpdatedTimestamp ?? null);
    filter.category.equalsOrIn(options.category ?? null);
    filter.externalId.prefix(options.externalIdPrefix ?? null);
    filter.space.equalsOrIn(options.space ?? null);

    return filter;
  }

  /**
   * Retrieve ProductNode instances by ID.
   *
   * @param id - Instance identifier. Can be a string, InstanceId, tuple, or array
   * @param options - Additional options
   * @returns For single id: The ProductNode if found, undefined otherwise.
   *          For array of ids: A ProductNodeList of found instances.
   */
  async retrieve(
    id: string | InstanceId | readonly [string, string],
    options?: { space?: string },
  ): Promise<ProductNode | undefined>;
  async retrieve(
    id: readonly (string | InstanceId | readonly [string, string])[],
    options?: { space?: string },
  ): Promise<ProductNodeList>;
  async retrieve(
    id:
      | string
      | InstanceId
      | readonly [string, string]
      | readonly (string | InstanceId | readonly [string, string])[],
    options: { space?: string } = {},
  ): Promise<ProductNode | ProductNodeList | undefined> {
    const isSingle = !Array.isArray(id) ||
      (id.length === 2 && typeof id[0] === "string" && typeof id[1] === "string");

    if (isSingle) {
      return await this._retrieve(
        id as string | InstanceId | readonly [string, string],
        options,
      ) as ProductNode | undefined;
    }
    const result = await this._retrieve(
      id as readonly (string | InstanceId | readonly [string, string])[],
      options,
    );
    return new ProductNodeList([...(result as InstanceList<ProductNode>)]);
  }

  /**
   * Iterate over instances with pagination.
   *
   * @param options - Filter and pagination options
   * @returns A Page containing items and optional next cursor.
   */
  async iterate(options: {
    name?: string | readonly string[];
    namePrefix?: string;
    description?: string | readonly string[];
    descriptionPrefix?: string;
    minPrice?: number;
    maxPrice?: number;
    minQuantity?: number;
    maxQuantity?: number;
    active?: boolean;
    minCreatedDate?: Date | string;
    maxCreatedDate?: Date | string;
    minUpdatedTimestamp?: Date | string;
    maxUpdatedTimestamp?: Date | string;
    category?: string | InstanceId | readonly [string, string] | readonly (string | InstanceId | readonly [string, string])[];
    externalIdPrefix?: string;
    space?: string | readonly string[];
    cursor?: string;
    limit?: number;
  } = {}): Promise<Page<ProductNodeList>> {
    const filter = this._buildFilter(options);
    const page = await this._iterate({
      cursor: options.cursor,
      limit: options.limit,
      filter: filter.asFilter(),
    });

    return { ...page, items: new ProductNodeList([...page.items]) };
  }

  /**
   * Search instances using full-text search.
   *
   * @param options - Search and filter options
   * @returns A ProductNodeList of matching instances.
   */
  async search(options: {
    query?: string;
    properties?: string | readonly string[];
    name?: string | readonly string[];
    namePrefix?: string;
    description?: string | readonly string[];
    descriptionPrefix?: string;
    minPrice?: number;
    maxPrice?: number;
    minQuantity?: number;
    maxQuantity?: number;
    active?: boolean;
    minCreatedDate?: Date | string;
    maxCreatedDate?: Date | string;
    minUpdatedTimestamp?: Date | string;
    maxUpdatedTimestamp?: Date | string;
    category?: string | InstanceId | readonly [string, string] | readonly (string | InstanceId | readonly [string, string])[];
    externalIdPrefix?: string;
    space?: string | readonly string[];
    limit?: number;
  } = {}): Promise<ProductNodeList> {
    const filter = this._buildFilter(options);

    const result = await this._search({
      query: options.query,
      properties: options.properties,
      limit: options.limit,
      filter: filter.asFilter(),
    });

    return new ProductNodeList([...result.items]);
  }

  /**
   * Aggregate instances.
   *
   * @param aggregate - Aggregation(s) to perform.
   * @param options - Filter and grouping options.
   * @returns AggregateResponse with aggregated values.
   */
  async aggregate(
    aggregate: Aggregation | readonly Aggregation[],
    options: {
      groupBy?: string | readonly string[];
    name?: string | readonly string[];
    namePrefix?: string;
    description?: string | readonly string[];
    descriptionPrefix?: string;
    minPrice?: number;
    maxPrice?: number;
    minQuantity?: number;
    maxQuantity?: number;
    active?: boolean;
    minCreatedDate?: Date | string;
    maxCreatedDate?: Date | string;
    minUpdatedTimestamp?: Date | string;
    maxUpdatedTimestamp?: Date | string;
    category?: string | InstanceId | readonly [string, string] | readonly (string | InstanceId | readonly [string, string])[];
    externalIdPrefix?: string;
    space?: string | readonly string[];
    } = {},
  ): Promise<AggregateResponse> {
    const filter = this._buildFilter(options);

    return this._aggregate(aggregate, {
      groupBy: options.groupBy,
      filter: filter.asFilter(),
    });
  }

  /**
   * List instances with type-safe filtering.
   *
   * @param options - Filter, sort, and pagination options.
   * @returns A ProductNodeList of matching instances.
   */
  async list(options: {
    name?: string | readonly string[];
    namePrefix?: string;
    description?: string | readonly string[];
    descriptionPrefix?: string;
    minPrice?: number;
    maxPrice?: number;
    minQuantity?: number;
    maxQuantity?: number;
    active?: boolean;
    minCreatedDate?: Date | string;
    maxCreatedDate?: Date | string;
    minUpdatedTimestamp?: Date | string;
    maxUpdatedTimestamp?: Date | string;
    category?: string | InstanceId | readonly [string, string] | readonly (string | InstanceId | readonly [string, string])[];
    externalIdPrefix?: string;
    space?: string | readonly string[];
    sortBy?: string;
    sortDirection?: SortDirection;
    limit?: number;
  } = {}): Promise<ProductNodeList> {
    const filter = this._buildFilter(options);
    const sort: PropertySort | undefined = options.sortBy !== undefined
      ? {
        property: createPropertyRef(PRODUCT_NODE_VIEW, options.sortBy),
        direction: options.sortDirection,
      }
      : undefined;

    const result = await this._list({
      limit: options.limit,
      filter: filter.asFilter(),
      sort,
    });

    return new ProductNodeList([...result]);
  }
}

File: examples_v2/example_sdk_python/_api/_product_node_api.py
from collections.abc import Sequence
from datetime import date, datetime
from typing import Literal, overload

from cognite.pygen._python.instance_api._api import InstanceAPI
from cognite.pygen._python.instance_api.http_client import HTTPClient
from cognite.pygen._python.instance_api.models import (
    Aggregation,
    InstanceId,
    PropertySort,
    ViewReference,
)
from cognite.pygen._python.instance_api.models.responses import (
    AggregateResponse,
    Page,
)

from ._data_class import (
    ProductNode,
    ProductNodeFilter,
    ProductNodeList,
)


def _create_property_ref(view_ref: ViewReference, property_name: str) -> list[str]:
    """Create a property reference for filtering."""
    return [view_ref.space, f"{view_ref.external_id}/{view_ref.version}", property_name]


class ProductNodeApi(InstanceAPI[ProductNode, ProductNodeList]):
    """API for ProductNode instances with type-safe filter methods."""

    def __init__(self, http_client: HTTPClient) -> None:
        view_ref = ViewReference(space="pygen_example", external_id="ProductNode", version="v1")
        super().__init__(http_client, view_ref, "node", ProductNodeList)

    @overload
    def retrieve(
        self,
        id: str | InstanceId | tuple[str, str],
        space: str | None = None,
    ) -> ProductNode | None: ...

    @overload
    def retrieve(
        self,
        id: list[str | InstanceId | tuple[str, str]],
        space: str | None = None,
    ) -> ProductNodeList: ...

    def retrieve(
        self,
        id: str | InstanceId | tuple[str, str] | list[str | InstanceId | tuple[str, str]],
        space: str | None = None,
    ) -> ProductNode | ProductNodeList | None:
        """Retrieve ProductNode instances by ID.

        Args:
            id: Instance identifier(s). Can be a string, InstanceId, tuple, or list of these.
            space: Default space to use when id is a string.

        Returns:
            For single id: The ProductNode if found, None otherwise.
            For list of ids: A ProductNodeList of found instances.
        """
        return self._retrieve(id, space)

    def aggregate(
        self,
        aggregate: Aggregation | Sequence[Aggregation],
        group_by: str | Sequence[str] | None = None,
        name: str | list[str] | None = None,
        name_prefix: str | None = None,
        description: str | list[str] | None = None,
        description_prefix: str | None = None,
        min_price: float | None = None,
        max_price: float | None = None,
        min_quantity: int | None = None,
        max_quantity: int | None = None,
        active: bool | None = None,
        min_created_date: date | None = None,
        max_created_date: date | None = None,
        min_updated_timestamp: datetime | None = None,
        max_updated_timestamp: datetime | None = None,
        category: str | InstanceId | tuple[str, str] | list[str | InstanceId | tuple[str, str]] | None = None,
        external_id_prefix: str | None = None,
        space: str | list[str] | None = None,
    ) -> AggregateResponse:
        """Aggregate instances.

        Args:
            aggregate: Aggregation(s) to perform.
            group_by: Property or properties to group by.
            name: Filter by exact name or list of values.
            name_prefix: Filter by name prefix.
            description: Filter by exact description or list of values.
            description_prefix: Filter by description prefix.
            min_price: Minimum price (inclusive).
            max_price: Maximum price (inclusive).
            min_quantity: Minimum quantity (inclusive).
            max_quantity: Maximum quantity (inclusive).
            active: Filter by active.
            min_created_date: Minimum created date (inclusive).
            max_created_date: Maximum created date (inclusive).
            min_updated_timestamp: Minimum updated timestamp (inclusive).
            max_updated_timestamp: Maximum updated timestamp (inclusive).
            category: Filter by category relation.
            external_id_prefix: Filter by external ID prefix.
            space: Filter by space.

        Returns:
            AggregateResponse with aggregated values.
        """
        filter_ = ProductNodeFilter("and")
        filter_.name.equals_or_in(name)
        filter_.name.prefix(name_prefix)
        filter_.description.equals_or_in(description)
        filter_.description.prefix(description_prefix)
        filter_.price.greater_than_or_equals(min_price)
        filter_.price.less_than_or_equals(max_price)
        filter_.quantity.greater_than_or_equals(min_quantity)
        filter_.quantity.less_than_or_equals(max_quantity)
        filter_.active.equals(active)
        filter_.created_date.greater_than_or_equals(min_created_date)
        filter_.created_date.less_than_or_equals(max_created_date)
        filter_.updated_timestamp.greater_than_or_equals(min_updated_timestamp)
        filter_.updated_timestamp.less_than_or_equals(max_updated_timestamp)
        filter_.category.equals_or_in(category)
        filter_.external_id.prefix(external_id_prefix)
        filter_.space.equals_or_in(space)
        return self._aggregate(aggregate=aggregate, group_by=group_by, filter=filter_.as_filter())

    def search(
        self,
        query: str | None = None,
        properties: str | Sequence[str] | None = None,
        name: str | list[str] | None = None,
        name_prefix: str | None = None,
        description: str | list[str] | None = None,
        description_prefix: str | None = None,
        min_price: float | None = None,
        max_price: float | None = None,
        min_quantity: int | None = None,
        max_quantity: int | None = None,
        active: bool | None = None,
        min_created_date: date | None = None,
        max_created_date: date | None = None,
        min_updated_timestamp: datetime | None = None,
        max_updated_timestamp: datetime | None = None,
        category: str | InstanceId | tuple[str, str] | list[str | InstanceId | tuple[str, str]] | None = None,
        external_id_prefix: str | None = None,
        space: str | list[str] | None = None,
        limit: int = 25,
    ) -> ProductNodeList:
        """Search instances using full-text search.

        Args:
            query: The search query string.
            properties: Properties to search in. If None, searches all text properties.
            name: Filter by exact name or list of values.
            name_prefix: Filter by name prefix.
            description: Filter by exact description or list of values.
            description_prefix: Filter by description prefix.
            min_price: Minimum price (inclusive).
            max_price: Maximum price (inclusive).
            min_quantity: Minimum quantity (inclusive).
            max_quantity: Maximum quantity (inclusive).
            active: Filter by active.
            min_created_date: Minimum created date (inclusive).
            max_created_date: Maximum created date (inclusive).
            min_updated_timestamp: Minimum updated timestamp (inclusive).
            max_updated_timestamp: Maximum updated timestamp (inclusive).
            category: Filter by category relation.
            external_id_prefix: Filter by external ID prefix.
            space: Filter by space.
            limit: Maximum number of results.

        Returns:
            A ProductNodeList with matching instances.
        """
        filter_ = ProductNodeFilter("and")
        filter_.name.equals_or_in(name)
        filter_.name.prefix(name_prefix)
        filter_.description.equals_or_in(description)
        filter_.description.prefix(description_prefix)
        filter_.price.greater_than_or_equals(min_price)
        filter_.price.less_than_or_equals(max_price)
        filter_.quantity.greater_than_or_equals(min_quantity)
        filter_.quantity.less_than_or_equals(max_quantity)
        filter_.active.equals(active)
        filter_.created_date.greater_than_or_equals(min_created_date)
        filter_.created_date.less_than_or_equals(max_created_date)
        filter_.updated_timestamp.greater_than_or_equals(min_updated_timestamp)
        filter_.updated_timestamp.less_than_or_equals(max_updated_timestamp)
        filter_.category.equals_or_in(category)
        filter_.external_id.prefix(external_id_prefix)
        filter_.space.equals_or_in(space)
        return self._search(query=query, properties=properties, limit=limit, filter=filter_.as_filter()).items

    def iterate(
        self,
        name: str | list[str] | None = None,
        name_prefix: str | None = None,
        description: str | list[str] | None = None,
        description_prefix: str | None = None,
        min_price: float | None = None,
        max_price: float | None = None,
        min_quantity: int | None = None,
        max_quantity: int | None = None,
        active: bool | None = None,
        min_created_date: date | None = None,
        max_created_date: date | None = None,
        min_updated_timestamp: datetime | None = None,
        max_updated_timestamp: datetime | None = None,
        category: str | InstanceId | tuple[str, str] | list[str | InstanceId | tuple[str, str]] | None = None,
        external_id_prefix: str | None = None,
        space: str | list[str] | None = None,
        cursor: str | None = None,
        limit: int = 25,
    ) -> Page[ProductNodeList]:
        """Iterate over instances with pagination.

        Args:
            name: Filter by exact name or list of values.
            name_prefix: Filter by name prefix.
            description: Filter by exact description or list of values.
            description_prefix: Filter by description prefix.
            min_price: Minimum price (inclusive).
            max_price: Maximum price (inclusive).
            min_quantity: Minimum quantity (inclusive).
            max_quantity: Maximum quantity (inclusive).
            active: Filter by active.
            min_created_date: Minimum created date (inclusive).
            max_created_date: Maximum created date (inclusive).
            min_updated_timestamp: Minimum updated timestamp (inclusive).
            max_updated_timestamp: Maximum updated timestamp (inclusive).
            category: Filter by category relation.
            external_id_prefix: Filter by external ID prefix.
            space: Filter by space.
            cursor: Pagination cursor from a previous page.
            limit: Maximum number of results per page (1-1000).

        Returns:
            A Page containing items and optional next cursor.
        """
        filter_ = ProductNodeFilter("and")
        filter_.name.equals_or_in(name)
        filter_.name.prefix(name_prefix)
        filter_.description.equals_or_in(description)
        filter_.description.prefix(description_prefix)
        filter_.price.greater_than_or_equals(min_price)
        filter_.price.less_than_or_equals(max_price)
        filter_.quantity.greater_than_or_equals(min_quantity)
        filter_.quantity.less_than_or_equals(max_quantity)
        filter_.active.equals(active)
        filter_.created_date.greater_than_or_equals(min_created_date)
        filter_.created_date.less_than_or_equals(max_created_date)
        filter_.updated_timestamp.greater_than_or_equals(min_updated_timestamp)
        filter_.updated_timestamp.less_than_or_equals(max_updated_timestamp)
        filter_.category.equals_or_in(category)
        filter_.external_id.prefix(external_id_prefix)
        filter_.space.equals_or_in(space)
        return self._iterate(cursor=cursor, limit=limit, filter=filter_.as_filter())

    def list(
        self,
        name: str | list[str] | None = None,
        name_prefix: str | None = None,
        description: str | list[str] | None = None,
        description_prefix: str | None = None,
        min_price: float | None = None,
        max_price: float | None = None,
        min_quantity: int | None = None,
        max_quantity: int | None = None,
        active: bool | None = None,
        min_created_date: date | None = None,
        max_created_date: date | None = None,
        min_updated_timestamp: datetime | None = None,
        max_updated_timestamp: datetime | None = None,
        category: str | InstanceId | tuple[str, str] | list[str | InstanceId | tuple[str, str]] | None = None,
        external_id_prefix: str | None = None,
        space: str | list[str] | None = None,
        sort_by: str | None = None,
        sort_direction: Literal["ascending", "descending"] = "ascending",
        limit: int | None = 25,
    ) -> ProductNodeList:
        """List instances with type-safe filtering.

        Args:
            name: Filter by exact name or list of values.
            name_prefix: Filter by name prefix.
            description: Filter by exact description or list of values.
            description_prefix: Filter by description prefix.
            min_price: Minimum price (inclusive).
            max_price: Maximum price (inclusive).
            min_quantity: Minimum quantity (inclusive).
            max_quantity: Maximum quantity (inclusive).
            active: Filter by active.
            min_created_date: Minimum created date (inclusive).
            max_created_date: Maximum created date (inclusive).
            min_updated_timestamp: Minimum updated timestamp (inclusive).
            max_updated_timestamp: Maximum updated timestamp (inclusive).
            category: Filter by category relation.
            external_id_prefix: Filter by external ID prefix.
            space: Filter by space.
            sort_by: Property name to sort by.
            sort_direction: Sort direction.
            limit: Maximum number of results. None for no limit.

        Returns:
            A ProductNodeList of matching instances.
        """
        filter_ = ProductNodeFilter("and")
        filter_.name.equals_or_in(name)
        filter_.name.prefix(name_prefix)
        filter_.description.equals_or_in(description)
        filter_.description.prefix(description_prefix)
        filter_.price.greater_than_or_equals(min_price)
        filter_.price.less_than_or_equals(max_price)
        filter_.quantity.greater_than_or_equals(min_quantity)
        filter_.quantity.less_than_or_equals(max_quantity)
        filter_.active.equals(active)
        filter_.created_date.greater_than_or_equals(min_created_date)
        filter_.created_date.less_than_or_equals(max_created_date)
        filter_.updated_timestamp.greater_than_or_equals(min_updated_timestamp)
        filter_.updated_timestamp.less_than_or_equals(max_updated_timestamp)
        filter_.category.equals_or_in(category)
        filter_.external_id.prefix(external_id_prefix)
        filter_.space.equals_or_in(space)
        sort = None
        if sort_by is not None:
            prop_ref = _create_property_ref(self._view_ref, sort_by)
            sort = PropertySort(property=prop_ref, direction=sort_direction)

        return self._list(limit=limit, filter=filter_.as_filter(), sort=sort)
